DigitalLearningSolutions.Web/Controllers/NotificationPreferencesController.cs

```diff
@@ -61,11 +68,17 @@ public IActionResult UpdateNotificationPreferences(string userType)
         [Authorize]
         [HttpPost]
         [Route("/NotificationPreferences/Edit/{userType}")]
-        public IActionResult SaveNotificationPreferences(string userType, IEnumerable<int> notificationIds)
+        public IActionResult SaveNotificationPreferences(UserType userType, IEnumerable<int> notificationIds)
         {
-            var userId = ((UserType)userType).Equals(UserType.AdminUser)
+            var userId = Equals(userType, UserType.AdminUser)
                 ? User.GetAdminId()
-                : User.GetCandidateIdKnownNotNull();
+                : Equals(userType, UserType.DelegateUser)
+                    ? User.GetCandidateId()
+                    : null;
+            if (userId == null)
+            {
+                return NotFound();
+            }

```

Comment:
I find it's a bit difficult to read multiple ternary statements chained together.
Is there a reason you've not gone for a switch statement?
That would allow you to put the `return NotFound()` closer to the condition in which it gets run too.
Same on other endpoint

Reply:
From my understanding a switch statement needs constant values to compare against; we're using object instances since UserType is a subclass of Enumeration not an actual enum. So I don't think it's possible (please enlighten if I'm wrong). But I've refactored and used if statements for clarity.

Reply:
It looks like they made some improvements in C# 7.0 (we're using 8.0): https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch#the-match-expression

I'd say it's worth giving it a go if you haven't already. If it doesn't work, what you've got looks great.

Reply:
The match expression can be anything but the case label patterns have to be constants or types. I'm afraid I still don't see how to get that to work.

Reply:
Ah, sorry, yep, I was barking up the wrong tree there. Fair enough!